packages/firestore/src/model/mutation.ts

```diff
@@ -226,13 +226,17 @@ export abstract class Mutation {
    *
    * @param maybeDoc The document to mutate. The input document can be null if
    *     the client has no knowledge of the pre-mutation state of the document.
+   * @param baseDoc The state of the document prior to this mutation batch. The
+   *     input document can be null if the client has no knowledge of the
+   *     pre-mutation state of the document.
```

Comment:
`set({a:FieldValue.serverTimestamp()})` is a good example why this doesn't work. It will create two mutations: One that clears the map `a` and removes the existing value, and one that adds the timestamp. When we apply the timestamp, `previousValue` has already been cleared in `maybeDoc`.